rfc/20250317-ephemeral-resources.md

```diff
@@ -0,0 +1,646 @@
+* lifecycle
+* count
+* for_each
+* depends_on
+* provider
+
+#### `Open` method details
+When OpenTofu will have to use an ephemeral resource, it needs to call its `Open` method, passing over the config of the ephemeral resource.
+
+The call to the `Open` method will return the following data:
+* `Private` that OpenTofu is not going to use in other contexts than calling the provider `Close` or `Renew` optionally defined methods.
+* `Result` that will contain the actual ephemeral information. This is what OpenTofu needs to handle to make it available to other ephemeral contexts to reference.
+* `RenewAt` being an optional timestamp indicating when OpenTofu will have to call `Renew` method on the provider before using again the data from the `Result`.
+
+Observations:
+* In the `Result`, OpenTofu is epecting to find any non-computed given values in the request, otherwise will return an error.
+* In the `Result`, the fields marked as computed can be either null or have an actual value. If an unknown if found, OpenTofu will return an error.
+
+> [!NOTE]
+>
+> If any information in the configuration of an ephemeral resource is unknown during the `plan` phase, OpenTofu will defer the provisioning of the resource for the `apply` phase.
+
+#### `Renew` method details
+The `Renew` method is called only if the response from `Open` or another `Renew` call is containing a `RenewAt`.
+When `RenewAt` is present, OpenTofu, before using the `Result` from the `Open` method response, will check if the current timestamp is at or over `RenewAt` and will call the `Renew` method by providing the previously returned `Private` information.
+
+> [!NOTE]
+> 
+> `Renew` does not return a *new* information meant to replace the initial `Result` returned by the `Open` call.
+> Due to this, `Renew` is only useful for system similar to Vault where the lease can be renewed without generating new data.
+
+#### `Close` method details
+When OpenTofu is done using an ephemeral resource, it needs to call its `Close` method to ensure that any remote data associated with the data returned in `OpenResponse.Result` is released and/or cleaned up properly.
+The `Renew` request is requiring the latest `Private` data returned by the call to `Open` or `Renew` method.
+
+#### `ConfigValidators` and `ValidateConfig` methods details
+There is not much to say here, since this is the same lifecycle that a datasource is having.
+
+#### Testing support
+The testing support will be documented later into a different RFC, or as amendment to this one.
+
+### Support in already ephemeral contexts
+There are already OpenTofu contexts that are not saved in state/plan file:
+* `provider` configuration
+* `provisioner` blocks
+* `connection` blocks
+
+In all of these, referencing an ephemeral value should work as normal.
+
+### Utilities
+#### `terraform.applying`
+The `terraform.applying` needs to be introduced to allow the user to check if the current command that is running is `apply` or not.
```

Comment:
why terraform and not tofu?

Reply:
Because of the feature parity. Maybe having both just to be able to migrate? And in case we encounter `terraform.applying` write a warning that we also support `tofu.applying`? 🤔

Reply:
Internally, they both map to the same structure for now.  Let's at least mention tofu.applying here.

Reply:
Done e8c189287f194d9ea8f2319658605ad22d618f45